util/notes/warnings.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notes
+
+import (
+	"errors"
+	"fmt"
+)
+
+type Warnings []error
+
+func (ws Warnings) Merge(notes Notes) Warnings {
+	nws := append(ws, notes.Warnings...)
+	for _, a := range notes.Annotations {
+		nws = append(nws, errors.New(a))
+	}
+	return nws
+}
+
+//nolint:revive // Ignore ST1012
+var (
+	RangeTooSmallWarning         = errors.New("Need at least 2 points to compute, perhaps time range is too small")
+	MixedFloatsHistogramsWarning = errors.New("Range contains a mix of histograms and floats")
+	MixedOldNewHistogramsWarning = errors.New("Range contains a mix of conventional and native histograms")
+
+	InvalidQuantileWarning    = errors.New("Quantile value should be between 0 and 1")
+	BadBucketLabelWarning     = errors.New("No bucket label or malformed label value")
+	PossibleNonCounterWarning = errors.New("Metric might not be a counter (name does not end in _total/_sum/_count)")
+)
+
+func IsForEmptyResultOnly(err error) bool {
+	return errors.Is(err, RangeTooSmallWarning)
```

Comment:
I see what you are trying here, but I'm afraid it won't really solve the problem:
* We will still see false positives if the selected vector happens to have only one time series, and that series legitimately has just started (so that the range may be ever so long, it would not see another sample).
* This will definitely create false negatives, for example if there are many series in the vector and the user has picked a range close to the scrape interval. In that case, some series might succeed to yield a rate, while most won't.